pkg/controllermanager/controller/shoot/maintenance/reconciler.go

```diff
@@ -93,44 +96,53 @@ func requeueAfterDuration(shoot *gardencorev1beta1.Shoot) (time.Duration, time.T
 	return duration, nextMaintenance
 }
 
-func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gardencorev1beta1.Shoot, gardenClient client.Client) error {
+func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gardencorev1beta1.Shoot) error {
 	log.Info("Maintaining Shoot")
 
+	var (
+		maintainedShoot = shoot.DeepCopy()
+		operations      []string
+	)
+
 	cloudProfile := &gardencorev1beta1.CloudProfile{}
 	if err := r.Client.Get(ctx, kutil.Key(shoot.Spec.CloudProfileName), cloudProfile); err != nil {
 		return err
 	}
 
-	reasonForImageUpdatePerPool, err := MaintainMachineImages(log, shoot, cloudProfile)
+	reasonForImageUpdatePerPool, err := maintainMachineImages(log, maintainedShoot, cloudProfile)
 	if err != nil {
 		// continue execution to allow the kubernetes version update
 		log.Error(err, "Failed to maintain Shoot machine images")
 	}
+	operations = append(operations, reasonForImageUpdatePerPool...)
 
-	reasonForKubernetesUpdate, err := maintainKubernetesVersion(log, shoot.Spec.Kubernetes.Version, shoot.Spec.Maintenance.AutoUpdate.KubernetesVersion, cloudProfile, func(v string) error {
-		shoot.Spec.Kubernetes.Version = v
+	reasonForKubernetesUpdate, err := maintainKubernetesVersion(log, maintainedShoot.Spec.Kubernetes.Version, maintainedShoot.Spec.Maintenance.AutoUpdate.KubernetesVersion, cloudProfile, func(v string) error {
+		maintainedShoot.Spec.Kubernetes.Version = v
 		return nil
 	})
 	if err != nil {
 		// continue execution to allow the machine image version update
 		log.Error(err, "Failed to maintain Shoot kubernetes version")
 	}
+	if reasonForKubernetesUpdate != "" {
```

Comment:
You don't have this `if` condition above for `reasonForImageUpdatePerPool`. Why here? Can we harmonize this?

Reply:
There a list is returned so if the list is empty nothing will be appended but here if I don't check if the string is nil i.e. shoot Kubernetes version was not upgraded, it will append an empty string to the list.